src/gui/editors/AutomationEditor.cpp

```diff
@@ -492,162 +400,228 @@ void AutomationEditor::drawLine( int x0In, float y0, int x1In, float y1 )
+						m_drawLastTick = posTicks;
+						m_drawLastLevel = level;
 
-			// left button??
-			if( mouseEvent->button() == Qt::LeftButton &&
-							m_editMode == DRAW )
-			{
-				m_pattern->addJournalCheckPoint();
-				// Connect the dots
-				if( mouseEvent->modifiers() & Qt::ShiftModifier )
-				{
-					drawLine( m_drawLastTick,
-							m_drawLastLevel,
-							pos_ticks, level );
-					m_mouseDownLeft = false;
-				}
-				m_drawLastTick = pos_ticks;
-				m_drawLastLevel = level;
+						// Changes the action to drawing a line of nodes
+						m_action = DRAW_LINE;
+					}
+					else // No shift, we are just creating/moving nodes
+					{
+						// If the place we clicked had no nodes
+						if (clickedNode == tm.end())
+						{
+							// We create a new node and start dragging it
+							TimePos valuePos(posTicks);
+
+							// Starts actually moving/draging the node
+							TimePos newTime = m_pattern->setDragValue(
+								valuePos,
+								level,
+								true,
+								mouseEvent->modifiers() & Qt::ControlModifier
+							);
+
+							// Set the iterator to our newly created node so we can use it later
+							// for the m_moveXOffset calculation
+							clickedNode = tm.find(newTime);
+						}
+						else // If we clicked over an existing node
+						{
+							// Simply start moving/draging it
+							TimePos newTime = m_pattern->setDragValue(
+								TimePos(POS(clickedNode)),
+								level,
+								true,
+								mouseEvent->modifiers() & Qt::ControlModifier
+							);
+
+							// We need to update our iterator because setDragValue removes the node that
+							// is being dragged, so if we don't update it we have a bogus iterator
+							clickedNode = tm.find(newTime);
+						}
 
-				// did it reach end of map because
-				// there's no value??
-				if( it == time_map.end() )
-				{
-					// then set new value
-					TimePos value_pos( pos_ticks );
-
-					TimePos new_time =
-						m_pattern->setDragValue( value_pos,
-									level, true,
-							mouseEvent->modifiers() &
-								Qt::ControlModifier );
-
-					// reset it so that it can be used for
-					// ops (move, resize) after this
-					// code-block
-					it = time_map.find( new_time );
+						// Set the action to MOVE_VALUE so moveMouseEvent() knows we are moving a node
+						m_action = MOVE_VALUE;
+
+						// Calculate the offset from the place the mouse click happened in comparison
+						// to the center of the node
+						int alignedX = (POS(clickedNode) - m_currentPosition) * m_ppb / TimePos::ticksPerBar();
+						m_moveXOffset = x - alignedX - 1;
+
+						// Set move-cursor
+						QCursor c(Qt::SizeAllCursor);
+						QApplication::setOverrideCursor(c);
+
+						// Update the last clicked position so it can be used if we draw a line later
+						m_drawLastTick = posTicks;
+						m_drawLastLevel = level;
+					}
+
+					Engine::getSong()->setModified();
 				}
+				else if (m_mouseDownRight) // Right click on DRAW mode erases values
```

Comment:
We have some code duplications in this functions. Is it worth pulling them into some lambda functions?

Reply:
This also applies to mouse drag events.

Reply:
I think it might be one of those cases where aiming for fewer code lines is costly to the clarity of the logic. Definitely there are some statements that could still be extracted to outside the `if` blocks, but then we are moving them out of the context in which they are used and anyone reading the code for the first time will have to make an extra effort figuring it out.

Of course this is a personal opinion and if it's disagreed with I can still move the code to avoid repetition, but I think the way it is organized might take more lines of code but also makes it much easier to understand the code and implement new functionality. Are there particular duplications of code you had in mind? Maybe they are not even amongst the ones that benefit from the context and I can just move them :p

Reply:
I agree with making some lambda's in order to reduce the duplication. More duplication means larger surface for bugs to creep in, or if making a change a lot of more places to fix.

Reply:
Applies your requested changes @Veratil ! And apologies, it took me re-reading the comments to realize what was being suggested. For some reason I misunderstood lambda for ternary and had the wrong idea in mind. You mean extracting the "Remove" and "Reset" nodes subroutines to something like this?
```
auto eraseNode = [this](timeMap::iterator node)
{
    if (node != tm.end())
    {
        m_pattern->removeNode(POS(node));
        Engine::getSong()->setModified();
    }
    m_action = ERASE_VALUES;
}
```

Then using `eraseNode(clickedNode);`?

Reply:
Right, except for maybe the `m_action = ERASE_VALUES;` part. We have a lot of the pattern: `if node != end then removeNode, setModified endif` which can be pulled out to a lambda. Same with maybe a few other things.

Reply:
Would you check if the last commit is what you two meant?

Reply:
That's it! We'll see if @PhysSong sees any more that can be done.